handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -836,98 +869,76 @@ public final SSLEngineResult unwrap(
-                        }
-                    } else {
-                        int sslError = SSL.getError(ssl, bytesRead);
-                        switch (sslError) {
-                            case SSL.SSL_ERROR_ZERO_RETURN:
-                                // This means the connection was shutdown correctly, close inbound and outbound
-                                if (!receivedShutdown) {
-                                    closeAll();
-                                }
-                                // fall-trough!
-                            case SSL.SSL_ERROR_WANT_READ:
-                            case SSL.SSL_ERROR_WANT_WRITE:
-                                // break to the outer loop
-                                return newResult(isInboundDone() ? CLOSED : OK, bytesConsumed, bytesProduced, status);
-                            default:
-                                return sslReadErrorResult(SSL.getLastErrorNumber(), bytesConsumed, bytesProduced);
+                            int sslError = SSL.getError(ssl, bytesRead);
+                            switch (sslError) {
+                                case SSL.SSL_ERROR_WANT_WRITE:
+                                case SSL.SSL_ERROR_WANT_READ:
+                                    // break to the outer loop as we want to read more data which means we need to write
+                                    // more to the BIO.
+                                    break readLoop;
+
+                                case SSL.SSL_ERROR_ZERO_RETURN:
+                                    // This means the connection was shutdown correctly, close inbound and outbound
+                                    if (!receivedShutdown) {
+                                        closeAll();
+                                    }
+                                    return newResult(
+                                            isInboundDone() ? CLOSED : OK, bytesConsumed, bytesProduced, status);
+
+                                default:
+                                    return sslReadErrorResult(SSL.getLastErrorNumber(), bytesConsumed, bytesProduced);
+                            }
                         }
                     }
-                }
-            } else {
-                // If the capacity of all destination buffers is 0 we need to trigger a SSL_read anyway to ensure
-                // everything is flushed in the BIO pair and so we can detect it in the pendingAppData() call.
-                if (SSL.readFromSSL(ssl, EMPTY_ADDR, 0) <= 0) {
-                    // We do not check SSL_get_error as we are not interested in any error that is not fatal.
-                    int err = SSL.getLastErrorNumber();
-                    if (OpenSsl.isError(err)) {
-                        return sslReadErrorResult(err, bytesConsumed, bytesProduced);
-                    }
-                }
-            }
-            if (pendingAppData() > 0) {
```

Comment:
can you explain why this was removed? is it no longer possible for the encryption portion to generate more data than can fit in the destination buffers? the `capacity` vs `packetLength` we check above is against the encrypted length not the decrypted length, right? Is the extra iteration over `dsts` even necessary if we can just deal with the `BUFFER_OVERFLOW` at the end instead of trying to check upfront based off of encrypted length?